### PR TITLE
Create ArrayList classe

### PR DESCRIPTION
**Pull request by inbox-reva**
Hi! Please check my solution.

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,48 +1,135 @@
...
+                    return i;
+                }
+            }
+        } else {
+            for (int i = 0; i < size; i++) {
+                if (element.equals(this.elements[i])) {
+                    return i;
+                }
+            }
+        }
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:132]: remove this outer if-else

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,48 +1,135 @@
...
+    private void grow(int minCapacity) {
+        int oldCapacity = this.elements.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        if (newCapacity - minCapacity < 0) {
+            newCapacity = minCapacity;
+        }
+        if (newCapacity > Integer.MAX_VALUE - 8) {
+            throw new OutOfMemoryError();
+        }
+        this.elements = Arrays.copyOf(this.elements, newCapacity);
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:116]: use system.arraycopy

**Review by aaananas — CHANGES_REQUESTED** [review]
Before you create PR with your homework, you need to go through the checklist under the task and correct all the points described there. The mentor will not check the work until the checklist points are corrected

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,48 +1,132 @@
...
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Illegal Index: " + index);
+        }
+        ensureCapacity(size + 1);
+        System.arraycopy(this.elements, index, this.elements, index + 1, size - index);
+        this.elements[index] = value;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:33]:
```suggestion
        elements[index] = value;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,48 +1,132 @@
...
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Illegal Index: " + index);
+        }
+        ensureCapacity(size + 1);
+        System.arraycopy(this.elements, index, this.elements, index + 1, size - index);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:32]: Delete all 'this' in all methods

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,48 +1,132 @@
...
+        this.elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int numNew = list.size();
+        T[] arrayList = (T[]) new Object[numNew];
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:40]: numNew, arrayList - what is it? 
image

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,48 +1,132 @@
...
+        System.arraycopy(arrayList, 0, elements, size, numNew);
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Illegal Index: " + index);
+        }
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/ArrayList.java:52]: image

**Review by Rommelua — CHANGES_REQUESTED** [review]
https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/array-list.html

Before you create PR with your homework, you need to go through the checklist under the task and correct all the points described there. The mentor will not check the work until the checklist points are corrected

File: src/main/java/core/basesyntax/ArrayList.java (hunk 6 above)
Comment (Rommelua) [reply src/main/java/core/basesyntax/ArrayList.java:52]: not fixed

**Review by Rommelua — CHANGES_REQUESTED** [review]
You should fix (or answer to) all mentor's comments before re-request new review. Be attentive and do not hesitate to contact us in Slack if you do not understand something.

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,48 +1,133 @@
...
+            if (element == null && elements[i] == null) {
+                return i;
+            }
+        }
+        for (int i = 0; i < size; i++) {
+            if (element.equals(elements[i])) {
+                return i;
+            }
+        }
+        return -1;
```
Comment (aaananas) [inline src/main/java/core/basesyntax/ArrayList.java:124]:
```suggestion
        for (int i = 0; i < size; i++) {
            if (element == elements[i] || element != null && element.equals(elements[i]) {
                return i;
            }
        }
        return -1;
```

**Review by aaananas — APPROVED** [review]

**Pull request closed without merging**